Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,218 @@
+            }
+
+        case .running:
+            throw ServiceRunnerError.alreadyRunning(file: file, line: line)
+
+        case .finished:
+            throw ServiceRunnerError.alreadyFinished(file: file, line: line)
+        }
+    }
+
+    private func _run() async throws {
+        self.logger.info(
+            "Starting service lifecycle",
+            metadata: [
+                self.configuration.loggingConfiguration.signalsKey: "\(self.configuration.gracefulShutdownSignals)",
+                self.configuration.loggingConfiguration.servicesKey: "\(self.services)",
+            ]
+        )
+
+        enum ChildTaskResult {
+            case serviceFinished(any Service)
+            case serviceThrew(any Service, any Error)
+            case signalCaught(UnixSignal)
+            case signalSequenceFinished
+        }
+
+        let shutdownGracefullyManager = GracefulShutdownManager()
+        try await TaskLocals.$gracefulShutdownManager.withValue(shutdownGracefullyManager) {
+            try await withThrowingTaskGroup(of: ChildTaskResult.self) { group in
+                // First we have to register our signals.
+                let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+
+                group.addTask {
+                    for await signal in unixSignals {
+                        return .signalCaught(signal)
+                    }
+
+                    return .signalSequenceFinished
+                }
+
+                for service in self.services {
+                    self.logger.debug(
+                        "Starting service",
+                        metadata: [
+                            self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                        ]
+                    )
+
+                    group.addTask {
+                        do {
+                            try await service.run()
+                            return .serviceFinished(service)
+                        } catch {
+                            return .serviceThrew(service, error)
+                        }
+                    }
+                }
+
+                var finishedServiceCount = 0
+
+                // We are going to wait for any of the services to finish or
+                // the signal sequence to throw an error.
+                while !group.isEmpty {
+                    // No child task is actually throwing here so the try! is safe
+                    let result = try! await group.next()
+
+                    switch result {
+                    case .serviceFinished(let service):
+                        if service.isLongRunning {
+                            // If a long running service finishes early we treat this as an unexpected
+                            // early exit and have to cancel the rest of the services.
+                            self.logger.error(
+                                "Service finished unexpectedly",
```

Comment:
Probably worth mentioning that it's a long-running service to hint back to the API